hy/compiler.py

```diff
@@ -1850,24 +1850,41 @@ def compile_for_expression(self, expression):
     @checkargs(min=2)
     def compile_while_expression(self, expr):
         expr.pop(0)  # "while"
-        ret = self.compile(expr.pop(0))
+        cond = expr.pop(0)
+        cond_compiled = self.compile(cond)
 
-        orel = Result()
-        # (while cond body (else …))
+        else_expr = None
         if expr and expr[-1][0] == HySymbol("else"):
             else_expr = expr.pop()
+
+        if cond_compiled.stmts:
+            # We need to ensure the statements for the condition are
+            # executed on every iteration. Rewrite the loop to use a
+            # single anonymous variable as the condition.
+            def e(*x): return HyExpression(x)
+            s = HySymbol
+            cond_var = s(self.get_anon_var())
+            return self.compile(e(
+                s('do'),
+                e(s('setv'), cond_var, s('True')),
+                e(s('while'), cond_var,
+                  e(s('setv'), cond_var, e(s('not'), e(s('not'), cond))),
```

Comment:
IMO for simplicity purposes (also because `(not (not ...))` is pointless), this should be compiled to something like this instead:

```hy
(while True
  (setv anon-var expr-here)
  (if-not anon-var (break)))
```

Right now the condition is being tested twice, and again the double nots aren't necessary.

Reply:
@kirbyfan64 The problem with that design is that `break` skips its loop's `else` clause. As for the not-not, that's for the case in which `cond` produces a mutable value and its truth value changes. See the tests in this PR and the discussion in #1432.

Reply:
@Kodiologist touche. Couldn't you still do `anon-var.__bool__()` though?

Reply:
@kirbyfan64 That doesn't seem to exist in Python 2.

Reply:
It's `__nonzero__` on Python 2.

Reply:
@kirbyfan64 Not every class defines `__nonzero__`.

Reply:
FWIW I'd at least prefer a comment explaining the `not not `.

Reply:
@kirbyfan64 Done.